### PR TITLE
add mesh public doc-strings

### PR DESCRIPTION
**Pull request by bjlittle**
### Description
This PR adds the doc-strings to the public API methods of the class `iris.experimental.ugrid.Mesh` et al.

To view the rendered doc-strings for this pull-request, see [here](https://bjlittle-iris.readthedocs.io/en/latest/generated/api/iris/experimental/ugrid.html).

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

### CODE REVIEW

File: lib/iris/experimental/ugrid.py [hunk 1]
```diff
@@ -1345,6 +1679,68 @@ def remove_connectivities(
...
+
+        * attributes (dict):
+            A dictionary of attributes desired on the coordinates. If ``None``,
+            does not check for ``attributes``.
+
+        * axis (str):
+            The desired coordinate axis, see :func:`~iris.util.guess_coord_axis`.
+            If ``None``, does not check for ``axis``. Accepts the values ``X``,
+            ``Y``, ``Z`` and ``T`` (case-insensitive).
+
```
Comment (trexfeathers) [inline lib/iris/experimental/ugrid.py:1723]: No `axis` for this one.

But it **does** need `cf_role` instead.

File: lib/iris/experimental/ugrid.py [hunk 2]
```diff
@@ -1414,17 +1883,67 @@ def xml_element(self):
     #     # use Connectivity.indices_by_src for fetching indices, passing in the lazy_indices() result as an argument.
 
     def dimension_names_reset(self, node=False, edge=False, face=False):
+        """
+        Clear the current assigned name to be used for the netCDF variable
+        representing the ``node``, ``edge`` and ``face`` dimension to
+        ``None``.
```
Comment (trexfeathers) [inline lib/iris/experimental/ugrid.py:1889]:
```suggestion
        Reset the name(s) used for the netCDF variable(s)
        representing the ``node``, ``edge`` and/or ``face`` dimension(s)
        to their default values.
```

File: lib/iris/experimental/ugrid.py [hunk 3]
```diff
@@ -1414,17 +1883,67 @@ def xml_element(self):
...
+
+        Kwargs:
+
+        * node (bool):
+            Clear the name of the ``node`` dimension if ``True``. Default
```
Comment (trexfeathers) [inline lib/iris/experimental/ugrid.py:1894]:
```suggestion
            Reset the name of the ``node`` dimension if ``True``. Default
```

File: lib/iris/experimental/ugrid.py [hunk 4]
```diff
@@ -1414,17 +1883,67 @@ def xml_element(self):
...
+            is ``False``.
+
+        * edge (bool):
+            Clear the name of the ``edge`` dimension if ``True``. Default
```
Comment (trexfeathers) [inline lib/iris/experimental/ugrid.py:1898]:
```suggestion
            Reset the name of the ``edge`` dimension if ``True``. Default
```

File: lib/iris/experimental/ugrid.py [hunk 5]
```diff
@@ -1414,17 +1883,67 @@ def xml_element(self):
...
+            is ``False``.
+
+        * face (bool):
+            Clear the name of the ``face`` dimension if ``True``. Default
```
Comment (trexfeathers) [inline lib/iris/experimental/ugrid.py:1902]:
```suggestion
            Reset the name of the ``face`` dimension if ``True``. Default
```

File: lib/iris/experimental/ugrid.py [hunk 6]
```diff
@@ -1414,17 +1883,67 @@ def xml_element(self):
...
+            is ``False``.
+
+        """
         return self._set_dimension_names(node, edge, face, reset=True)
 
     def dimension_names(self, node=None, edge=None, face=None):
+        """
+        Assign the name to be used for the netCDF variable representing
+        the ``node``, ``edge`` and ``face`` dimension.
```
Comment (trexfeathers) [inline lib/iris/experimental/ugrid.py:1911]:
```suggestion
        Assign the name(s) to be used for the netCDF variable(s) representing
        the ``node``, ``edge`` and ``face`` dimension(s).
```

**Review by trexfeathers — CHANGES_REQUESTED** [review]
Thanks @bjlittle!

These docstrings are clear, informative and appropriately mirror phrasing in the [UGRID conventions](https://ugrid-conventions.github.io/ugrid-conventions/).

I've picked up on some corrections that are needed.

As mentioned in specific comments on the code, I have trouble with the level of repetition between the filter methods, their singular equivalents and the remove methods. As I read/scroll through the rendered docs this repetition makes it much harder to keep my place. I like the pattern currently used for documenting `Cube.coord()` as a solution.

File: lib/iris/experimental/ugrid.py [hunk 7]
```diff
@@ -168,7 +188,7 @@ def __init__(
         * start_index (int):
             Either ``0`` or ``1``. Default is ``0``. Denotes whether
             :attr:`indices` uses 0-based or 1-based indexing (allows support
-            for Fortran and legacy NetCDF files).
+            for Fortran and legacy netCDF files).
```
[the remark this answers is not in the document]
Comment (bjlittle) [reply lib/iris/experimental/ugrid.py:191]: It's refered inconsistently through out Iris, typically I refer to it as `NetCDF`, but corrected it to the most common usage in the file... think I might stick to that, as it is a bugbear...

File: lib/iris/experimental/ugrid.py [hunk 8]
```diff
@@ -1116,12 +1116,47 @@ def _set_dimension_names(self, node, edge, face, reset=False):
 
         return result
 
+    @property
+    def all_connectivities(self):
+        """
+        All the :class:`Connectivity` coordinates of the :class:`Mesh`.
```
[the remark this answers is not in the document]
Comment (bjlittle) [reply lib/iris/experimental/ugrid.py:1122]: Me too, so I changed this :+1:

[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/experimental/ugrid.py:1122]: I don't follow; it still says "`Connectivity` **coordinates**"? I don't mind if we keep it this way, but it sounds like you're saying you changed it? Am I missing some of your changes?

File: lib/iris/experimental/ugrid.py [hunk 9]
```diff
@@ -1273,6 +1425,71 @@ def connectivity(
         contains_edge=None,
         contains_face=None,
     ):
+        """
+        Return a single :class:`Connectivity` coordinate from the :class:`Mesh` that
+        matches the provided criteria.
+
+        Criteria can be either specific properties or other objects with
+        metadata to be matched.
+
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/experimental/ugrid.py:1440]: I'm inferring from the change that you are "definitely not in favour", as I put it. Could you do me the courtesy of explaining your reasoning? I see the current state as significantly difficult to read and maintain, but I'm assuming there's a good reason behind it?

[the remark this answers is not in the document]
Comment (bjlittle) [reply lib/iris/experimental/ugrid.py:1440]: Yes, if as a user I click a link I want all the information I need, not a reference to another link.

We originally done that as a convenience to ourselves as maintainers and not the user. The docs are user centric, so being explicit is better.

[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/experimental/ugrid.py:1440]: OK that makes sense. I had been viewing this as a user reading an RTD webpage but of course that's only half the story. I certainly get as frustrated as anyone with Matplotlib's ubiquitous stub docstrings with `*args, **kwargs`!

File: lib/iris/experimental/ugrid.py [hunk 10]
```diff
@@ -60,27 +63,44 @@
...
+#: Namedtuple for ``edge`` :class:`~iris.coords.AuxCoord` coordinates.
 MeshEdgeCoords = namedtuple("MeshEdgeCoords", ["edge_x", "edge_y"])
+#: Namedtuple for ``face`` :class:`~iris.coords.AuxCoord` coordinates.
 MeshFaceCoords = namedtuple("MeshFaceCoords", ["face_x", "face_y"])
 
+#
 # Mesh connectivity manager namedtuples.
+#
+
+#: Namedtuple for 1D mesh :class:`Connectivity` coordinates.
```
Comment (trexfeathers) [inline lib/iris/experimental/ugrid.py:101]:
```suggestion
#: Namedtuple for 1D mesh :class:`Connectivity` instances.
```

File: lib/iris/experimental/ugrid.py [hunk 11]
```diff
@@ -60,27 +63,44 @@
...
 Mesh1DConnectivities = namedtuple("Mesh1DConnectivities", ["edge_node"])
+#: Namedtuple for 2D mesh :class:`Connectivity` coordinates.
```
Comment (trexfeathers) [inline lib/iris/experimental/ugrid.py:103]:
```suggestion
#: Namedtuple for 2D mesh :class:`Connectivity` instances.
```

File: lib/iris/experimental/ugrid.py [hunk 12]
```diff
@@ -1249,6 +1346,67 @@ def connectivities(
...
+              :attr:`~iris.common.mixin.CFVariableMixin.long_name`, or
+              :attr:`~iris.common.mixin.CFVariableMixin.var_name` which is
+              compared against the :meth:`~iris.common.mixin.CFVariableMixin.name`.
+
+            * a connectivity or metadata instance equal to that of
+              the desired objects e.g., :class:`Connectivity` or
+              :class:`ConnectivityMetadata`.
+
+        * standard_name (str):
+            The CF standard name of the desired :class:`Connectibity`. If
```
Comment (trexfeathers) [inline lib/iris/experimental/ugrid.py:1375]:
```suggestion
            The CF standard name of the desired :class:`Connectivity`. If
```

File: lib/iris/experimental/ugrid.py [hunk 13]
```diff
@@ -1345,6 +1696,66 @@ def remove_connectivities(
...
         contains_face=None,
     ):
+        """
+        Remove one or more :class:`Connectivity` from the :class:`Mesh` that
+        match the provided criteria.
+
+        Criteria can be either specific properties or other objects with
+        metadata to be matched.
+
+        Kwargs:
```
Comment (trexfeathers) [inline lib/iris/experimental/ugrid.py:1706]: Need to remove the word `coordinate` by replicating the Kwarg descriptions you used for `Mesh.connectivities()` and `Mesh.connectivity()`.

File: lib/iris/experimental/ugrid.py [hunk 14]
```diff
@@ -1273,6 +1431,74 @@ def connectivity(
...
+
+        .. note::
+
+            If the given criteria do not return **precisely one**
+            :class:`Connectivity`, then a
+            :class:`~iris.exceptions.ConnectivityNotFoundError` is raised.
+
+        .. seealso::
+
+            :meth:`Cube.connectivities` for matching zero or more connectivities.
```
Comment (trexfeathers) [inline lib/iris/experimental/ugrid.py:1449]:
```suggestion
            :meth:`Mesh.connectivities` for matching zero or more connectivities.
```

File: lib/iris/experimental/ugrid.py [hunk 15]
```diff
@@ -1297,6 +1523,71 @@ def coord(
...
+        metadata to be matched.
+
+        .. note::
+
+            If the given criteria do not return **precisely one** coordinate,
+            then a :class:`~iris.exceptions.CoordinateNotFoundError` is raised.
+
+        .. seealso::
+
+            :meth:`Cube.coords` for matching zero or more coordinates.
```
Comment (trexfeathers) [inline lib/iris/experimental/ugrid.py:1540]:
```suggestion
            :meth:`Mesh.coords` for matching zero or more coordinates.
```

**Review by trexfeathers — CHANGES_REQUESTED** [review]
Thanks for the latest clarifications @bjlittle. Here are some final requests in light of my improved understanding. Once these are addressed I'll be delighted to merge 😊

File: lib/iris/experimental/ugrid.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (bjlittle) [reply lib/iris/experimental/ugrid.py:1440]: It is true that some people read the API docs end-to-end... I must admit I do do that sometimes, and it makes for a pretty disjointed read, as normally users dip into the API docs to look at a single specific thing.

So yeah, I get your point that this is overly verbose, but generally I think that's actually okay.

**Comment by trexfeathers** [comment]
Amazing work as usual @bjlittle, LGTM

**Comment by bjlittle** [comment]
@trexfeathers Hallelujah :100: :rocket: :partying_face: 

Thanks for sticking with it :beers:

**Pull request merged**